pkg/volume/gce_pd/gce_util.go

```diff
@@ -108,33 +102,49 @@ func (util *GCEDiskUtil) DetachDisk(pd *gcePersistentDisk) error {
 	globalPDPath := makeGlobalPDName(pd.plugin.host, pd.pdName)
 	glog.V(5).Infof("DetachDisk(pd) where pd is %#v and the globalPDPath is %q\r\n", pd, globalPDPath)
 
-	// Terminate any in progress verify detach go routines, this will block until the goroutine is ready to exit because the channel is unbuffered
-	detachCleanupManager.Send(pd.pdName, true)
-
 	if err := pd.mounter.Unmount(globalPDPath); err != nil {
 		return err
 	}
 	if err := os.Remove(globalPDPath); err != nil {
 		return err
 	}
-	// Detach the disk
-	gce, err := cloudprovider.GetCloudProvider("gce", nil)
-	if err != nil {
-		return err
-	}
-	if err := gce.(*gce_cloud.GCECloud).DetachDisk(pd.pdName); err != nil {
-		return err
+
+	if detachCleanupManager.Exists(pd.pdName) {
+		glog.Warningf("Terminating new DetachDisk call for GCE PD %q. A previous detach call for this PD is still pending.", pd.pdName)
+		return nil
+
 	}
 
-	// Verify disk detached, retry if needed.
-	go verifyDetached(pd, gce)
+	// Detach disk, retry if needed.
+	go detachDiskAndVerify(pd)
 	return nil
 }
 
-// Verifys the disk device to be created has been succesffully attached, and retries if it fails.
-func verifyAttached(pd *gcePersistentDisk, sdBeforeSet util.StringSet, gce cloudprovider.Interface) (string, error) {
+// Attaches the specified persistent disk device to node, verifies that it is attached, and retries if it fails.
+func attachDiskAndVerify(pd *gcePersistentDisk, sdBeforeSet util.StringSet) (string, error) {
 	devicePaths := getDiskByIdPaths(pd)
+	var gce cloudprovider.Interface
 	for numRetries := 0; numRetries < maxRetries; numRetries++ {
+		if gce == nil {
+			var err error
+			gce, err = cloudprovider.GetCloudProvider("gce", nil)
+			if err != nil {
+				// Retry on error. See issue #11321
+				glog.Errorf("Error getting GCECloudProvider while attaching PD %q: %v", pd.pdName, err)
+				time.Sleep(errorSleepDuration)
+				continue
+			}
+		}
+
+		if numRetries > 0 {
+			glog.Warningf("Timed out waiting for GCE PD %q to attach. Retrying attach.", pd.pdName)
+		}
+
+		if err := gce.(*gce_cloud.GCECloud).AttachDisk(pd.pdName, pd.readOnly); err != nil {
+			// Retry on error. See issue #11321
+			glog.Errorf("Error attaching PD %q: %v", pd.pdName, err)
```

Comment:
Actually I just remembered I did this on prupose. A previous attach/detach operation could succeed therefore it is worth checking at every iteration whether the disk is attached/detached, instead of doing nothing and sleeping, we can do verification--so I remeoved the sleep as well.